src/data_distribution/first_touch_distribution.hpp

```diff
@@ -41,107 +45,122 @@ namespace argo {
-			if (result == 0) {
-				homenode_bit_id = id;
+			/* check/try to acquire ownership of the page with CAS to process' cas_node index */
+			argo::backend::atomic::_compare_exchange_owners_dir(&rank, &global_null, &result, sizeof(std::size_t), cas_node, owners_dir_window_index+2);
+
+			/* this process was the first one to deposit its rank */
+			if (result == global_null) {
+				/* iterate through the nodes to find a valid offset for the page, starting from the currently running one */
+				std::size_t n, searched;
+				for(n = rank, searched = 0; searched < static_cast<std::size_t>(base_distribution<instance>::nodes);
+						n = (n + 1) % base_distribution<instance>::nodes, searched++) {
+					bool succeeded = false;
+					/* load backing offset for the node from the offsets table */
+					argo::backend::atomic::_load_local_offsets_tbl(&offset, rank, n);
+					/* check if the backing store size of this node is sufficient */
+					while (offset < base_distribution<instance>::size_per_node) {
+						/* try to claim a valid offset */
+						const std::size_t incr_offset = offset + granularity;
+						argo::backend::atomic::_compare_exchange_offsets_tbl(&incr_offset, &offset, &result, sizeof(std::size_t), n, n);
+						if (result == offset) { succeeded = true; homenode = n; break; }
+						offset = result;
+					}
+					/* cache the offset returned from a remote CAS operation */
+					if (n != rank) {
+						argo::backend::atomic::_store_local_offsets_tbl(offset, rank, n);
+					}
+					/* succeeded, leave */
+					if (succeeded) break;
+				}
 
 				/* mark the page in the local window */
-				argo::backend::atomic::_store_local_dir(id, argo::backend::node_id(), owner_window_index);
-				argo::backend::atomic::_store_local_dir(argo::backend::backing_offset(), argo::backend::node_id(), owner_window_index+1);
+				argo::backend::atomic::_store_local_owners_dir(homenode, rank, owners_dir_window_index);
+				argo::backend::atomic::_store_local_owners_dir(offset, rank, owners_dir_window_index+1);
+				argo::backend::atomic::_store_local_owners_dir(rank, rank, owners_dir_window_index+2);
 
 				/* mark the page in the public windows */
-				node_id_t n;
-				for(n = 0; n < base_distribution<instance>::nodes; n++) {
-					if (n != argo::backend::node_id()) {
-						argo::backend::atomic::_store_public_dir(&id, sizeof(std::size_t), n, owner_window_index);
-						argo::backend::atomic::_store_public_dir(&argo::backend::backing_offset(), sizeof(std::size_t), n, owner_window_index+1);
+				for(n = 0; n < static_cast<std::size_t>(base_distribution<instance>::nodes); n++) {
+					if (n != rank) {
+						argo::backend::atomic::_store_public_owners_dir(&homenode, sizeof(std::size_t), n, owners_dir_window_index);
+						argo::backend::atomic::_store_public_owners_dir(&offset, sizeof(std::size_t), n, owners_dir_window_index+1);
+						argo::backend::atomic::_store_public_owners_dir(&rank, sizeof(std::size_t), n, owners_dir_window_index+2);
```

Comment:
The one real concern I have with this approach is in this code block. This is essentially a broadcast of 3 values to _every_ node in the system, which will cost `3*(n-1)` remote operations for when each page is initially touched. With four nodes this may seem trivial, but with 64 nodes this is 189 remote operations the first time _each_ page is touched.

Each remote operation consists of the following sequence:
```
MPI_Win_lock(MPI_LOCK_EXCLUSIVE, ...);
MPI_Put(desired, 1, t_type, rank, disp, 1, t_type, owners_dir_window); // 1 value
MPI_Win_unlock(...)
```
The `Put` operation is guaranteed to have completed when `MPI_Win_unlock` returns, so there is no optimization to be had when placing three of them in succession. If instead, we write back all 3 values at once, since they are simply 3 successive integers, we reduce the amount of remote operations from `3*(n-1)` to simply `n-1`, as such:
```
/* put all 3 values in a buffer */
MPI_Win_lock(MPI_WIN_EXCLUSIVE, ...);
MPI_Put(buffer, 3, t_type, rank, disp, 3, t_type, owners_dir_window); // all 3 values
MPI_Win_unlock(...);
```
This still leaves expensive broadcasting in the system. I believe it would be significantly cheaper to write this to _one_ location, the so called `cas_node` you are already using for this page, and have each node query the `cas_node` the first time they load a page and only then add it to the local table. This would cost _at most_ the same number of remote operations as the modified broadcasting solution (one per node not performing the first touch = `n-1` remote operations), but they will be deferred to the point where the node actually wants the data. The worst-case cost is still the same, but this assumes that _every_ node touches _every_ page, and the average cost should be _significantly_ lower.

I suspect that this approach will overall perform much better in a first-touch scenario where data is not shared by multiple nodes, and that the cost of the current solution is hidden behind initialization that is not included in time measurements in our benchmarks.